File: test/hydra/cmds/stop_test.go
package cmds

import (
	"os"
	"runtime"
	"strings"
	"testing"
	"time"

	"github.com/micro-plat/hydra"
	"github.com/micro-plat/hydra/hydra/servers/http"
	"github.com/micro-plat/hydra/test/assert"
)

func Test_stop_Normal_running(t *testing.T) {
	resetServiceName(t.Name())
	execPrint(t)
	//正常
	defunc, fileCallback := injectStdOutFile()
	defer defunc()

	var app = hydra.NewApp(
		hydra.WithServerTypes(http.API),
		hydra.WithPlatName("xxtest"),
		hydra.WithSystemName("apiserver"),
		hydra.WithClusterName("c"),
	)
	hydra.Conf.API(":19020")

	//1. 删除服务
	os.Args = []string{"xxtest", "remove"}
	app.Start()
	time.Sleep(time.Second * 2)

	//2. 安装服务
	os.Args = []string{"xxtest", "install", "-r", runRegistryAddr, "-c", "c"}
	go app.Start()
	time.Sleep(time.Second * 2)

	//3. 启动服务
	os.Args = []string{"xxtest", "start"}
	go app.Start()

	time.Sleep(time.Second)

	//4. 关闭服务状态
	os.Args = []string{"xxtest", "stop"}
	go app.Start()
	time.Sleep(time.Second)

	//5. 清除服务
	os.Args = []string{"xxtest", "remove"}
	go app.Start()

	time.Sleep(time.Second * 2)

	bytes, err := fileCallback()

	if err != nil {
		t.Error(err)
		return
	}
	line := string(bytes)
	lines := strings.Split(line, "\r")
	for _, row := range lines {
		if strings.Contains(row, "Stopping") {
			line = row
			break
		}
	}

	if runtime.GOOS == "linux" || runtime.GOOS == "darwin" {
		result := strings.Contains(line, "sudo") || (strings.Contains(line, "Stopping") && strings.Contains(line, "OK"))
		assert.Equal(t, true, result, "关闭正常服务运行")
	}
	if runtime.GOOS == "windows" {
		result := (strings.Contains(line, "Stopping") && strings.Contains(line, "OK"))
		assert.Equal(t, true, result, "关闭正常服务运行")
	}
	time.Sleep(time.Second)
}

func Test_stop_Not_installed(t *testing.T) {
	//未安装的服务
	resetServiceName(t.Name())
	execPrint(t)
	defunc, fileCallback := injectStdOutFile()
	defer defunc()

	var app = hydra.NewApp(
		hydra.WithServerTypes(http.API),
		hydra.WithPlatName("xxtest"),
		hydra.WithSystemName("apiserver"),
		hydra.WithClusterName("c"),
	)
	hydra.Conf.API(":19010")

	//1. 清除服务(保证没有服务安装)
	os.Args = []string{"xxtest", "remove"}
	go app.Start()
	time.Sleep(time.Second * 2)

	//2. 启动服务
	os.Args = []string{"xxtest", "stop"}
	go app.Start()
	time.Sleep(time.Second * 2)

	bytes, err := fileCallback()
	if err != nil {
		t.Error(err)
		return
	}
	line := string(bytes)

	if runtime.GOOS == "linux" || runtime.GOOS == "darwin" {
		//unbuntu/centos
		result := strings.Contains(line, "sudo") || strings.Contains(line, "Service is not installed")
		assert.Equal(t, true, result, "停止未安装的服务")
	}
	if runtime.GOOS == "windows" {
		result := strings.Contains(line, "not exist as an installed service")
		assert.Equal(t, true, result, "停止未安装的服务")
	}

	time.Sleep(time.Second)
}

func Test_stop_has_stopped(t *testing.T) {
	//已停止的服务

	resetServiceName(t.Name())
	execPrint(t)

	defunc, fileCallback := injectStdOutFile()
	defer defunc()

	var app = hydra.NewApp(
		hydra.WithServerTypes(http.API),
		hydra.WithPlatName("xxtest"),
		hydra.WithSystemName("apiserver"),
		hydra.WithClusterName("c"),
	)
	hydra.Conf.API(":19010")

	//1. 清除服务(保证没有服务安装)
	os.Args = []string{"xxtest", "remove"}
	go app.Start()
	time.Sleep(time.Second * 2)

	//2. 安装服务
	os.Args = []string{"xxtest", "install", "-r", runRegistryAddr, "-c", "c"}
	go app.Start()
	time.Sleep(time.Second * 2)

	//3. 关闭服务状态
	os.Args = []string{"xxtest", "stop"}
	go app.Start()
	time.Sleep(time.Second)

	//4. 清理服务
	os.Args = []string{"xxtest", "remove"}
	go app.Start()
	time.Sleep(time.Second * 2)

	bytes, err := fileCallback()

	if err != nil {
		t.Error(err)
		return
	}
	line := string(bytes)

	if runtime.GOOS == "linux" || runtime.GOOS == "darwin" {
		//unbuntu/centos
		result := strings.Contains(line, "sudo") || strings.Contains(line, "has already been stopped")
		assert.Equal(t, true, result, "停止关闭的服务")
	}
	if runtime.GOOS == "windows" {
		result := strings.Contains(line, "The service has not been started")
		assert.Equal(t, true, result, "停止关闭的服务")
	}

	time.Sleep(time.Second)
}

File: test/hydra/cmds/stop_test.py
# -*- coding: utf-8 -*-  

import os ,platform
import test_all
from test_help import test,execute,runApp
from test_help import ZKAddress

testName = os.path.basename(__file__)

@test(testName)
def test_stop_Normal_running():
    try:
        #1.安装服务
        args = ["install","-r",ZKAddress,"-c","c"]
        response = runApp(args)
        #print("install",response)
        if not "OK" in response:
            return u"安装服务失败"

        #2.启动
        args = ["start"]
        response = runApp(args)
        #print("start",response)
        if not "OK" in response:
            return u"启动服务失败"

        #3.停止
        args = ["stop"]
        response = runApp(args)
        #print("stop",response)
        if not "OK" in response:
            return u"停止服务失败"

    except Exception as err :
        return "error:"+err.message
        
    finally:    
        #4.删除
        args = ["remove"]
        response = runApp(args)
        #print("remove",response)

        if not "OK" in response:
            return u"删除服务失败"




@test(testName)
def test_stop_Not_installed():
    #1. 清理服务，避免其他遗留存在服务
    args = ["remove"]
    runApp(args)

    #2.启动服务
    args = ["stop"]
    response = runApp(args)
    
    if not ("not" in response and "installed" in response):
        return u"未安装服务验证"



@test(testName)
def test_stop_has_stopped():
    try:
        args = ["remove"]
        runApp(args)
        
        #1.安装服务
        args = ["install","-r",ZKAddress,"-c","c"]
        runApp(args)
       
        #3.停止
        args = ["stop"]
        response = runApp(args)  
        if "service has been stopped" not in response:
            return u"停止关闭的服务"

    except Exception as err :
        return "error:"+err.message
        
    finally:    
        #4.删除
        print("finally:remove")
        args = ["remove"]
        runApp(args)


def main():
     execute(testName)
 
if __name__ == "__main__":
    main()